upgrade.go

```diff
@@ -0,0 +1,231 @@
+package upgrade
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"runtime"
+	"time"
+
+	"github.com/blang/semver"
+	"github.com/pkg/errors"
+	"github.com/termie/go-shutil"
+)
+
+func CurrentVersion(version string) (semver.Version, error) {
+	return semver.Make(version)
+}
+
+func AvailableVersion(repo string, filePrefix string, channel string) (semver.Version, error) {
+	resp, err := http.Get(
+		fmt.Sprintf("%s/%s%s", repo, filePrefix, channel),
+	)
+	if err != nil {
+		return semver.Version{}, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return semver.Version{}, fmt.Errorf(
+			"while getting version; unexpected status %s",
+			resp.Status,
+		)
+	}
+	versionBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return semver.Version{}, err
+	}
+
+	return semver.Make(string(versionBytes))
+}
+
+// CheckAndNotifyIfOutOfDate checks if this utility is out of date compared to
+// the repository. It uses a caching strategy to avoid costly
+// checks with every run.
+func CheckAndNotifyIfOutOfDate(tool string, currentVersion string, repo string, filePrefix string, versionStable string, hoursToCheckForUpdate float64, upgradeCommand string) (bool, error) {
+	fn := filepath.Join(GetHome(), "."+tool+"-version-check")
+	file, err := os.OpenFile(fn, os.O_CREATE|os.O_RDWR, 0644)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "error: cannot create file (%s) to cache version checks\n", fn)
+		fmt.Fprintf(os.Stderr, "error: %s\n", err.Error())
+		return false, err
+	}
+	defer file.Close()
+
+	fi, err := file.Stat()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "error: cannot stat (%s)\n", fn)
+		fmt.Fprintf(os.Stderr, "error: %s\n", err.Error())
+		return false, err
+	}
+
+	diff := time.Now().Sub(fi.ModTime())
+	if diff.Hours() < hoursToCheckForUpdate {
+		return false, nil
+	}
+
+	// update mtime
+	err = file.Truncate(0)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "error: cannot update mtime (%s)\n", fn)
+		fmt.Fprintf(os.Stderr, "error: %s\n", err.Error())
+		return false, err
+	}
+
+	curr, err := CurrentVersion(currentVersion)
+	if err != nil {
+		return false, err
+	}
+	avail, err := AvailableVersion(repo, filePrefix, versionStable)
+	if err != nil {
+		return false, err
+	}
+
+	if avail.GT(curr) {
+		fmt.Fprintf(os.Stderr, "A new version of [%s] is available, run: %s\n", tool, upgradeCommand)
```

Comment:
Do we need to require "upgradeCommand" to have a value?

Reply:
I think that can wait until the interface is improved.